# Predicting the Success of Prospects in the NHL

The goal of this notebook is to predict the statistics of prospects in the NHL in their rookie years using several methods.

# Step 1: Preprocess the Data

In [1]:
# import needed libraries
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O

Convieniently, Kaggle provides a large csv file on players and statistics dating back to 1963 (including draft information)

In [ ]:
df = pd.read_csv("/content/drive/MyDrive/Datasets/nhldraft.csv")

In [ ]:
df.head()

id  year  overall_pick                 team            player nationality  \
0   1  2022             1   Montreal Canadiens  Juraj Slafkovsky          SK   
1   2  2022             2    New Jersey Devils       Simon Nemec          SK   
2   3  2022             3      Arizona Coyotes      Logan Cooley          US   
3   4  2022             4       Seattle Kraken      Shane Wright          CA   
4   5  2022             5  Philadelphia Flyers   Cutter Gauthier          SE   

  position   age  to_year                           amateur_team  ...  points  \
0       LW  18.0      NaN                          TPS (Finland)  ...     NaN   
1        D  18.0      NaN                    HK Nitra (Slovakia)  ...     NaN   
2        C  18.0      NaN  USA U-18 Development Team (USDP/USHL)  ...     NaN   
3        C  18.0      NaN              Kingston Frontenacs (OHL)  ...     NaN   
4       LW  18.0      NaN  USA U-18 Development Team (USDP/USHL)  ...     NaN   

   plus_minus  penalties_minutes  goalie_games_played  goalie_wins  \
0         NaN                NaN                  NaN          NaN   
1         NaN                NaN                  NaN          NaN   
2         NaN                NaN                  NaN          NaN   
3         NaN                NaN                  NaN          NaN   
4         NaN                NaN                  NaN          NaN   

   goalie_losses  goalie_ties_overtime  save_percentage  \
0            NaN                   NaN              NaN   
1            NaN                   NaN              NaN   
2            NaN                   NaN              NaN   
3            NaN                   NaN              NaN   
4            NaN                   NaN              NaN   

   goals_against_average  point_shares  
0                    NaN           NaN  
1                    NaN           NaN  
2                    NaN           NaN  
3                    NaN           NaN  
4                    NaN           NaN  

[5 rows x 23 columns]

In [ ]:
# for this scenario, we do not need the later columns
df = df.iloc[:, :10]
df.head()

id  year  overall_pick                 team            player nationality  \
0   1  2022             1   Montreal Canadiens  Juraj Slafkovsky          SK   
1   2  2022             2    New Jersey Devils       Simon Nemec          SK   
2   3  2022             3      Arizona Coyotes      Logan Cooley          US   
3   4  2022             4       Seattle Kraken      Shane Wright          CA   
4   5  2022             5  Philadelphia Flyers   Cutter Gauthier          SE   

  position   age  to_year                           amateur_team  
0       LW  18.0      NaN                          TPS (Finland)  
1        D  18.0      NaN                    HK Nitra (Slovakia)  
2        C  18.0      NaN  USA U-18 Development Team (USDP/USHL)  
3        C  18.0      NaN              Kingston Frontenacs (OHL)  
4       LW  18.0      NaN  USA U-18 Development Team (USDP/USHL)

Next, I will replace the NaN with zeros because any NaN information (except with 'to_year') should be replaced with a zero if it is not specified.

In [ ]:
df.iloc[:, 10:] = df.iloc[:, 10:].fillna(0)
df.head()

id  year  overall_pick                 team            player nationality  \
0   1  2022             1   Montreal Canadiens  Juraj Slafkovsky          SK   
1   2  2022             2    New Jersey Devils       Simon Nemec          SK   
2   3  2022             3      Arizona Coyotes      Logan Cooley          US   
3   4  2022             4       Seattle Kraken      Shane Wright          CA   
4   5  2022             5  Philadelphia Flyers   Cutter Gauthier          SE   

  position   age  to_year                           amateur_team  
0       LW  18.0      NaN                          TPS (Finland)  
1        D  18.0      NaN                    HK Nitra (Slovakia)  
2        C  18.0      NaN  USA U-18 Development Team (USDP/USHL)  
3        C  18.0      NaN              Kingston Frontenacs (OHL)  
4       LW  18.0      NaN  USA U-18 Development Team (USDP/USHL)

Now we have to convert categorical data into numeric so the data can be regressed on. For players and goalies, the only non-numeric data is:
*     the team that drafted them (team)
*     their nationality
*     their position (obviously for goalies it has been taken care of already)
*     their amateur team

This also ensures that when the ANOVA F-test is performed, that we can include categorical data such as nationality and the drafted team. For the amateur team, the league will be extracted from the data and then label encoding will be used to convert the categorical data into numeric ones.

In [ ]:
# change from amateur team to amateur league
import re
count = 0
for string in df['amateur_team']:
  res = re.findall(r'\(.*?\)', string)
  df['amateur_team'][count] = res[0].replace('(', '').replace(')', '')
  count = count + 1
df.head()

<ipython-input-6-ecf6b6a2b8c7>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['amateur_team'][count] = res[0].replace('(', '').replace(')', '')


id  year  overall_pick                 team            player nationality  \
0   1  2022             1   Montreal Canadiens  Juraj Slafkovsky          SK   
1   2  2022             2    New Jersey Devils       Simon Nemec          SK   
2   3  2022             3      Arizona Coyotes      Logan Cooley          US   
3   4  2022             4       Seattle Kraken      Shane Wright          CA   
4   5  2022             5  Philadelphia Flyers   Cutter Gauthier          SE   

  position   age  to_year amateur_team  
0       LW  18.0      NaN      Finland  
1        D  18.0      NaN     Slovakia  
2        C  18.0      NaN    USDP/USHL  
3        C  18.0      NaN          OHL  
4       LW  18.0      NaN    USDP/USHL

In [ ]:
df = df.rename(columns = {'amateur_team': 'amateur_league'})
df.amateur_league.value_counts()

OHL            1852
WHL            1721
QMJHL          1240
USHL            452
WCHL            413
               ... 
High-IL           1
Nova Scotia       1
Germany 2         1
RMJHL             1
GWMHA             1
Name: amateur_league, Length: 191, dtype: int64

With the teams, we can convert the teams into placements using the read_html() function from pandas.

In [ ]:
import time
# make a table to store the ranks of the teams
team_rank_tbl = pd.DataFrame()

#cycle through all the years from 1963-2022 for seasonal ranks
for year in range(1963,2023):
  #get the seasonal data from hockey-reference.com
  url = "https://www.hockey-reference.com/leagues/NHL_" + str(year) + "_standings.html"
  first_season = pd.DataFrame()
  #based on year, will have to take certain tables
  if year < 1968 or year==2021: #still original six and COVID year
    first_season = pd.read_html(url)[1]
  elif year == 2005: #lockout year
    continue
  else: # division into two conferences
    first_season = pd.read_html(url)[2]

  #format the dataframes to fit the criteria
  year_col = [year]*first_season.shape[0]
  first_season['year'] = year_col
  first_season.insert(1, "year", first_season.pop("year"))
  first_season.insert(2, "league_standing", first_season.pop("Rk"))
  first_season = first_season.iloc[:, 0:3]
  first_season = first_season.rename(columns = {"Unnamed: 1":"team"})

  #combine the new dataframe with the old one
  team_rank_tbl = pd.concat([first_season, team_rank_tbl], ignore_index=False, axis=0)

  #sleep to prevent HTTP Error 429 (Too Many Requests)
  time.sleep(3)
team_rank_tbl

year                 team  league_standing
0   2022     Florida Panthers                1
1   2022   Colorado Avalanche                2
2   2022  Carolina Hurricanes                3
3   2022  Toronto Maple Leafs                4
4   2022       Minnesota Wild                5
..   ...                  ...              ...
1   1963  Chicago Black Hawks                2
2   1963   Montreal Canadiens                3
3   1963    Detroit Red Wings                4
4   1963     New York Rangers                5
5   1963        Boston Bruins                6

[1334 rows x 3 columns]

Now we join the team rank table with the draft table to then add the team's league standing into the table.

In [ ]:
# join the two tables together
df = df.merge(team_rank_tbl, on=['team', 'year'])
df.insert(4, "league_standing", df.pop("league_standing"))
df.rename(columns={"league_standing":"drafted_team_standing"})

id  year  overall_pick                 team  drafted_team_standing  \
0          1  2022             1   Montreal Canadiens                     32   
1         26  2022            26   Montreal Canadiens                     32   
2         33  2022            33   Montreal Canadiens                     32   
3         62  2022            62   Montreal Canadiens                     32   
4         75  2022            75   Montreal Canadiens                     32   
...      ...   ...           ...                  ...                    ...   
11426  12249  1963            20     New York Rangers                      5   
11427  12235  1963             6  Toronto Maple Leafs                      1   
11428  12241  1963            12  Toronto Maple Leafs                      1   
11429  12246  1963            17  Toronto Maple Leafs                      1   
11430  12250  1963            21  Toronto Maple Leafs                      1   

                 player nationality position   age  to_year amateur_league  
0      Juraj Slafkovsky          SK       LW  18.0      NaN        Finland  
1           Filip Mesar          SK       RW  18.0      NaN       Slovakia  
2             Owen Beck          CA        C  18.0      NaN            OHL  
3           Lane Hutson          US        D  18.0      NaN      USDP/USHL  
4        Vinzenz Rohrer          AT        C  18.0      NaN            OHL  
...                 ...         ...      ...   ...      ...            ...  
11426  Campbell Alleson          CA        D   NaN      NaN                 
11427    Walt McKechnie          CA        C  16.0   1983.0         WOJBHL  
11428    Neil Clairmont          CA       LW   NaN      NaN                 
11429       Jim McKenny          CA        D  16.0   1979.0        MetJAHL  
11430      Gerry Meehan          CA        C  17.0   1979.0        MetJAHL  

[11431 rows x 11 columns]

For the amateur team, we can convert it into numerical data using Binary Encoding to allow every league of the 191 to be represented equally without drastically increasing the dimensionality of the dataset.

Binary Encoding assigns a number to the category, and then turns the number into its binary representation, making it very good to encode this feature, since it is a nominal feature (meaning there is no order between categories) and it has a lot of categories

In [ ]:
# convert amateur team using Binary Encoding
!pip install category_encoders
import category_encoders as ce

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 81.9/81.9 kB 2.0 MB/s eta 0:00:00


In [ ]:
#instantiate encoder (specifically BinaryEncoder)
encoder = ce.BinaryEncoder(cols=['amateur_league'])
#fit and transform
ama_team_binary = encoder.fit_transform(df['amateur_league'])
ama_team_binary
#merge it with the dataframe and remove the 'amateur_league' column
temp = df
temp = pd.concat([temp.iloc[:, :10], ama_team_binary], axis=1)
df = temp
df

id  year  overall_pick                 team  league_standing  \
0          1  2022             1   Montreal Canadiens               32   
1         26  2022            26   Montreal Canadiens               32   
2         33  2022            33   Montreal Canadiens               32   
3         62  2022            62   Montreal Canadiens               32   
4         75  2022            75   Montreal Canadiens               32   
...      ...   ...           ...                  ...              ...   
11426  12249  1963            20     New York Rangers                5   
11427  12235  1963             6  Toronto Maple Leafs                1   
11428  12241  1963            12  Toronto Maple Leafs                1   
11429  12246  1963            17  Toronto Maple Leafs                1   
11430  12250  1963            21  Toronto Maple Leafs                1   

                 player nationality position   age  to_year  amateur_league_0  \
0      Juraj Slafkovsky          SK       LW  18.0      NaN                 0   
1           Filip Mesar          SK       RW  18.0      NaN                 0   
2             Owen Beck          CA        C  18.0      NaN                 0   
3           Lane Hutson          US        D  18.0      NaN                 0   
4        Vinzenz Rohrer          AT        C  18.0      NaN                 0   
...                 ...         ...      ...   ...      ...               ...   
11426  Campbell Alleson          CA        D   NaN      NaN                 0   
11427    Walt McKechnie          CA        C  16.0   1983.0                 1   
11428    Neil Clairmont          CA       LW   NaN      NaN                 0   
11429       Jim McKenny          CA        D  16.0   1979.0                 0   
11430      Gerry Meehan          CA        C  17.0   1979.0                 0   

       amateur_league_1  amateur_league_2  amateur_league_3  amateur_league_4  \
0                     0                 0                 0                 0   
1                     0                 0                 0                 0   
2                     0                 0                 0                 0   
3                     0                 0                 0                 0   
4                     0                 0                 0                 0   
...                 ...               ...               ...               ...   
11426                 1                 0                 0                 1   
11427                 0                 0                 1                 0   
11428                 1                 0                 0                 1   
11429                 1                 1                 1                 0   
11430                 1                 1                 1                 0   

       amateur_league_5  amateur_league_6  amateur_league_7  
0                     0                 0                 1  
1                     0                 1                 0  
2                     0                 1                 1  
3                     1                 0                 0  
4                     0                 1                 1  
...                 ...               ...               ...  
11426                 0                 1                 0  
11427                 0                 1                 1  
11428                 0                 1                 0  
11429                 0                 1                 0  
11430                 0                 1                 0  

[11431 rows x 18 columns]

We can use the same process to convert the nationalities into numerical data

In [ ]:
# find the number of unique values in the nationalities column
print("number of distinct values: " + str(len(df.nationality.value_counts())))
df.nationality.value_counts()

number of distinct values: 46


CA    5998
US    2484
SE     758
RU     693
FI     472
CZ     450
SK     151
DE      80
CH      71
LV      35
UA      30
BY      29
DK      24
KZ      24
NO      23
GB      20
AT      19
FR       8
SI       8
PL       8
YU       3
UZ       3
HU       3
LT       3
JP       2
KR       2
BE       2
BS       2
BR       2
NG       2
EE       2
SU       2
ZA       1
HT       1
TW       1
PY       1
BN       1
AU       1
TZ       1
JM       1
ME       1
IT       1
NL       1
CN       1
TH       1
VE       1
Name: nationality, dtype: int64

In [ ]:
# convert the nationalities to numerical data using BinaryEncoder
country_encoder = ce.BinaryEncoder(cols=['nationality'])
#fit and transform
country_bin = country_encoder.fit_transform(df['nationality'])
country_bin

temp = df
temp = pd.concat([temp, country_bin], axis=1)
df = temp
df = df.drop('nationality', axis=1)
df

id  year  overall_pick                 team  league_standing  \
0          1  2022             1   Montreal Canadiens               32   
1         26  2022            26   Montreal Canadiens               32   
2         33  2022            33   Montreal Canadiens               32   
3         62  2022            62   Montreal Canadiens               32   
4         75  2022            75   Montreal Canadiens               32   
...      ...   ...           ...                  ...              ...   
11426  12249  1963            20     New York Rangers                5   
11427  12235  1963             6  Toronto Maple Leafs                1   
11428  12241  1963            12  Toronto Maple Leafs                1   
11429  12246  1963            17  Toronto Maple Leafs                1   
11430  12250  1963            21  Toronto Maple Leafs                1   

                 player position   age  to_year  amateur_league_0  ...  \
0      Juraj Slafkovsky       LW  18.0      NaN                 0  ...   
1           Filip Mesar       RW  18.0      NaN                 0  ...   
2             Owen Beck        C  18.0      NaN                 0  ...   
3           Lane Hutson        D  18.0      NaN                 0  ...   
4        Vinzenz Rohrer        C  18.0      NaN                 0  ...   
...                 ...      ...   ...      ...               ...  ...   
11426  Campbell Alleson        D   NaN      NaN                 0  ...   
11427    Walt McKechnie        C  16.0   1983.0                 1  ...   
11428    Neil Clairmont       LW   NaN      NaN                 0  ...   
11429       Jim McKenny        D  16.0   1979.0                 0  ...   
11430      Gerry Meehan        C  17.0   1979.0                 0  ...   

       amateur_league_4  amateur_league_5  amateur_league_6  amateur_league_7  \
0                     0                 0                 0                 1   
1                     0                 0                 1                 0   
2                     0                 0                 1                 1   
3                     0                 1                 0                 0   
4                     0                 0                 1                 1   
...                 ...               ...               ...               ...   
11426                 1                 0                 1                 0   
11427                 0                 0                 1                 1   
11428                 1                 0                 1                 0   
11429                 0                 0                 1                 0   
11430                 0                 0                 1                 0   

       nationality_0  nationality_1  nationality_2  nationality_3  \
0                  0              0              0              0   
1                  0              0              0              0   
2                  0              0              0              0   
3                  0              0              0              0   
4                  0              0              0              1   
...              ...            ...            ...            ...   
11426              0              0              0              0   
11427              0              0              0              0   
11428              0              0              0              0   
11429              0              0              0              0   
11430              0              0              0              0   

       nationality_4  nationality_5  
0                  0              1  
1                  0              1  
2                  1              0  
3                  1              1  
4                  0              0  
...              ...            ...  
11426              1              0  
11427              1              0  
11428              1              0  
11429              1              0  
11430       

Now, we split the data based upon draft year (taking players from the 2019 draft onward to test the models, using the 1963-2018 drafts to train the model)

In [ ]:
test_set = df.loc[df['year'] >= 2019]
test_set.tail()

id  year  overall_pick                   team  league_standing  \
869  872  2019           208        St. Louis Blues               12   
870  881  2019           217        St. Louis Blues               12   
871  768  2019           104  Columbus Blue Jackets               13   
872  778  2019           114  Columbus Blue Jackets               13   
873  876  2019           212  Columbus Blue Jackets               13   

               player position   age  to_year  amateur_league_0  ...  \
869    Vadim Zherenko        G  18.0      NaN                 0  ...   
870     Jeremy Michel       LW  18.0      NaN                 0  ...   
871       Eric Hjorth        D  18.0      NaN                 0  ...   
872  Dimitri Voronkov       LW  19.0      NaN                 0  ...   
873       Tyler Angle        C  18.0      NaN                 0  ...   

     amateur_league_4  amateur_league_5  amateur_league_6  amateur_league_7  \
869                 1                 0                 0                 1   
870                 1                 0                 0                 0   
871                 0                 1                 0                 1   
872                 0                 1                 1                 0   
873                 0                 0                 1                 1   

     nationality_0  nationality_1  nationality_2  nationality_3  \
869              0              0              0              1   
870              0              0              0              0   
871              0              0              0              1   
872              0              0              0              1   
873              0              0              0              0   

     nationality_4  nationality_5  
869              1              1  
870              1              0  
871              0              1  
872              1              1  
873              1              0  

[5 rows x 23 columns]

In [ ]:
train_set = df.loc[df['year'] < 2019]
train_set.head()

id  year  overall_pick            team  league_standing  \
874   882  2018             1  Buffalo Sabres               31   
875   913  2018            32  Buffalo Sabres               31   
876   975  2018            94  Buffalo Sabres               31   
877   998  2018           117  Buffalo Sabres               31   
878  1006  2018           125  Buffalo Sabres               31   

                        player position   age  to_year  amateur_league_0  ...  \
874              Rasmus Dahlin        D  18.0   2022.0                 0  ...   
875         Mattias Samuelsson        D  18.0   2022.0                 0  ...   
876                Matej Pekar        C  18.0      NaN                 0  ...   
877  Linus Lindstrand Cronholm        D  18.0      NaN                 0  ...   
878             Miska Kukkonen        D  18.0      NaN                 0  ...   

     amateur_league_4  amateur_league_5  amateur_league_6  amateur_league_7  \
874                 0                 1                 0                 0   
875                 0                 1                 0                 0   
876                 0                 1                 1                 1   
877                 0                 1                 0                 1   
878                 1                 1                 0                 0   

     nationality_0  nationality_1  nationality_2  nationality_3  \
874              0              0              0              1   
875              0              0              0              0   
876              0              0              1              0   
877              0              0              0              1   
878              0              0              0              1   

     nationality_4  nationality_5  
874              0              1  
875              1              1  
876              0              0  
877              0              1  
878              1              0  

[5 rows x 23 columns]

Additionally, we must split the players from the goalies.

In [ ]:
player_train_set = train_set.loc[train_set['position']!='G']
player_test_set = test_set.loc[test_set['position']!='G']

goalie_train_set = train_set.loc[train_set['position']=='G']
goalie_test_set = test_set.loc[test_set['position']=='G']

It is imperative before the dummy variables are created that we make sure that the position column does not have any typos or that equivalent positions but flipped order is equivalent.

In [ ]:
def change_pos_value(pos):
  if pos == "LW/C" or pos == "C; LW":
    return "C/LW"
  elif pos == "RW/C" or pos == "C RW" or pos == "C / R":
    return "C/RW"
  elif pos == "D/LW":
    return "LW/D"
  elif pos == "D/RW":
    return "RW/D"
  elif pos == "D/C":
    return "C/D"
  elif pos == "Centr":
    return "C"
  elif pos == "L/RW":
    return "W"
  elif pos == "C/W":
    return "F"
  elif pos == "D/W":
    return "W/D"
  else:
    return pos

position = player_train_set['position']
new_position = position.apply(change_pos_value)
player_train_set.position = new_position

test_pos = player_test_set['position']
new_test_pos = test_pos.apply(change_pos_value)
player_test_set.position = new_test_pos
player_test_set.position.value_counts()

<ipython-input-58-857eaa4e0d72>:25: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  player_train_set.position = new_position
<ipython-input-58-857eaa4e0d72>:29: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  player_test_set.position = new_test_pos


D     279
C     229
RW    149
LW    135
Name: position, dtype: int64

Since the number of positions are small, we can simply use one-hot encoding to represent the data in a way that can be fed into the machine learning algorithm. Obviously, this would only affect skaters since all goalies play the same position.

In [ ]:
# use one-hot encoding to create "dummy variables" for the SKATER positions
player_train_set = pd.get_dummies(player_train_set, columns=['position'])
player_test_set = pd.get_dummies(player_test_set, columns=['position'])

Now, we must combine these statistics with their respective rookie year statlines using Hockey Reference, the player's name, and their 'to_year' column in the table (since this excludes the 2022-2023 season, then the statistics only go from the 1962-63 season to the 2022-23 season), and this only applies to the training sets in both the skaters and the goalies.

In [ ]:
player_train_set

id  year  overall_pick                 team  league_standing  \
874      882  2018             1       Buffalo Sabres               31   
875      913  2018            32       Buffalo Sabres               31   
876      975  2018            94       Buffalo Sabres               31   
877      998  2018           117       Buffalo Sabres               31   
878     1006  2018           125       Buffalo Sabres               31   
...      ...   ...           ...                  ...              ...   
11426  12249  1963            20     New York Rangers                5   
11427  12235  1963             6  Toronto Maple Leafs                1   
11428  12241  1963            12  Toronto Maple Leafs                1   
11429  12246  1963            17  Toronto Maple Leafs                1   
11430  12250  1963            21  Toronto Maple Leafs                1   

                          player   age  to_year  amateur_league_0  \
874                Rasmus Dahlin  18.0   2022.0                 0   
875           Mattias Samuelsson  18.0   2022.0                 0   
876                  Matej Pekar  18.0      NaN                 0   
877    Linus Lindstrand Cronholm  18.0      NaN                 0   
878               Miska Kukkonen  18.0      NaN                 0   
...                          ...   ...      ...               ...   
11426           Campbell Alleson   NaN      NaN                 0   
11427             Walt McKechnie  16.0   1983.0                 1   
11428             Neil Clairmont   NaN      NaN                 0   
11429                Jim McKenny  16.0   1979.0                 0   
11430               Gerry Meehan  17.0   1979.0                 0   

       amateur_league_1  ...  position_C/LW  position_C/RW  position_D  \
874                   0  ...              0              0           1   
875                   0  ...              0              0           1   
876                   0  ...              0              0           0   
877                   0  ...              0              0           1   
878                   0  ...              0              0           1   
...                 ...  ...            ...            ...         ...   
11426                 1  ...              0              0           1   
11427                 0  ...              0              0           0   
11428                 1  ...              0              0           0   
11429                 1  ...              0              0           1   
11430                 1  ...              0              0           0   

       position_F  position_LW  position_LW/D  position_RW  position_RW/D  \
874             0            0              0            0              0   
875             0            0              0            0              0   
876             0            0              0            0              0   
877             0            0              0            0              0   
878             0            0              0            0              0   
...           ...          ...            ...          ...            ...   
11426           0            0              0            0              0   
11427           0            0              0            0              0   
11428           0            1              0            0              0   
11429           0            0              0            0              0   
11430           0            0              0            0              0   

       position_W  position_W/D  
874             0             0  
875             0             0  
876             0             0  
877             0             0  
878             0             0  
...           ...           ...  
11426           0             0  
11427           0             0  
11428           0             0  
11429           0             0  
11430           0             0  

[9495 rows x 34 columns]

In [32]:
import time
#merge the player statistics in the training set
final_pl_season_tbl = pd.DataFrame()

for year in range(1963, 2023):
  #access player stats during the seasonal data and format it with only
  #necessary data for the model
  if year == 2005:  #no player statistics for the 2005 lockout year
    continue
  url = "https://www.hockey-reference.com/leagues/NHL_" + str(year) + "_skaters.html"
  pl_season_tbl = pd.read_html(url)
  pl_season_tbl = pd.DataFrame(pl_season_tbl[0])
  pl_season_tbl = pl_season_tbl.iloc[:, 1:12]
  pl_season_tbl.columns = ['Player', 'Age', 'Tm', 'Pos', 'GP', 'G', 'A',
                          'PTS', '+/-', 'PIM', 'PS']
  pl_season_tbl = pl_season_tbl.drop_duplicates(subset='Player')
  pl_season_tbl.reset_index(drop=True, inplace=True)
  pl_season_tbl

  #concatenate tables as they come
  final_pl_season_tbl = pd.concat([pl_season_tbl, final_pl_season_tbl],
                                  ignore_index=True, axis=0)

  #remove duplicates, but keep last of the duplicates
  final_pl_season_tbl = final_pl_season_tbl.drop_duplicates(subset="Player", keep="last")

  #use this to prevent HTTP errors
  time.sleep(3)

final_pl_season_tbl

Player Age   Tm Pos  GP   G   A PTS  +/-  PIM   PS
0     Nicholas Abruzzese  22  TOR   C   9   1   0   1   -1    2  0.0
6     Alexander Alexeyev  22  WSH   D   1   0   0   0    0    2  0.0
26         Ronald Attard  22  PHI   D  15   2   2   4   -2    8  0.7
30       Brandon Baddock  26  MTL  LW   1   0   0   0    0    0  0.0
42         Justin Barron  20  TOT   D   7   1   1   2   -3    0  0.2
...                  ...  ..  ...  ..  ..  ..  ..  ..  ...  ...  ...
7149         Moose Vasko  27  CBH   D  64   4   9  13    6   70  4.8
7150         Ed Westfall  22  BOS   D  48   1  11  12   -8   34  1.4
7151       Kenny Wharram  29  CBH  RW  55  20  18  38   27   17  3.6
7152      Tommy Williams  22  BOS  RW  69  23  20  43  -13   11  3.1
7153         Howie Young  25  DET   D  64   4   5   9   -9  273  4.2

[6303 rows x 11 columns]

In [ ]:
#merge the goalie statistics in the training set

# Step 2: Finding Significance using ANOVA F-test

Before the models can be created, significance has to be tested among the variables.

For the players, the models will be predicting:

*   games played
*   goals
*   assists
*   points
*   penalty minutes
*   point shares (estimation of how many of team's points a player contributes to)


For the goalies, the models will be predicting:

*   games played
*   wins
*   save percentage
*   goals against average (GAA)

An ANOVA F-test can help determine correlations between the features of the train set and the variables that will be predicted on